File: jvm/tests/rtc/pcf8523/PCF8523Test.java
///usr/bin/env jbang "$0" "$@" ; exit $?
//JAVA 22+
//JAVA_OPTIONS --enable-native-access=ALL-UNNAMED
//DEPS it.uhde:periph-connection:1.2.0
//DEPS it.uhde:periph-java:1.2.0

import it.uhde.periph.connection.I2CConnection;
import it.uhde.periph.chips.rtc.PCF8523Minimal;
import it.uhde.periph.chips.rtc.PCF8523Full;

public class PCF8523Test {
    static int passed = 0, failed = 0;

    static void checkTrue(boolean cond, String label) {
        if (cond) { System.out.println("PASS " + label); passed++; }
        else      { System.out.println("FAIL " + label); failed++; }
    }

    public static void main(String[] args) throws Exception {
        int bus = Integer.parseInt(System.getenv().getOrDefault("I2C_BUS", "1"));

        try (var connection = new I2CConnection(bus, PCF8523Minimal.DEFAULT_ADDRESS)) {
            var rtc = new PCF8523Minimal(connection);
            checkTrue(true, "construct_minimal");

            rtc.setDatetime(2026, 9, 23, 3, 14, 30, 0);
            var dt = rtc.getDatetime();
            checkTrue(dt.year() == 2026 && dt.month() == 9 && dt.day() == 23 && dt.weekday() == 3, "datetime_roundtrip_date");
            checkTrue(dt.hour() == 14 && (dt.minute() == 30 || dt.minute() == 31), "datetime_roundtrip_time");

            var full = new PCF8523Full(connection);
            checkTrue(!full.oscillatorStopped(), "oscillator_running_after_set_datetime");

            var alarm = new PCF8523Full.Alarm(15, 6, null, null);
            full.setAlarm(alarm);
            checkTrue(alarm.equals(full.getAlarm()), "alarm_roundtrip");
            full.setAlarm(new PCF8523Full.Alarm(null, null, null, null));

            full.setOffset(-3, PCF8523Full.OffsetMode.EVERY_MINUTE);
            checkTrue(new PCF8523Full.Offset(-3, PCF8523Full.OffsetMode.EVERY_MINUTE).equals(full.getOffset()), "offset_roundtrip");
            full.setOffset(0);

            // Timer B at 64 Hz from 64 counts expires after ~1 s.
            full.pollInterrupt();
            full.configureTimerB(64, PCF8523Full.SourceClock.HZ_64);
            Thread.sleep(1500);
            checkTrue((full.pollInterrupt() & PCF8523Full.SOURCE_TIMER_B) != 0, "timer_b_fires");
            full.disableTimerB();
        }

        System.out.println("===DONE: " + passed + " passed, " + failed + " failed===");
        System.exit(failed == 0 ? 0 : 1);
    }
}

File: nodejs/tests/rtc/pcf8523_test.js
'use strict';

const { I2CConnection } = require('../../packages/periph/src/connection/i2c');
const { PCF8523Full } = require('../../packages/periph/src/chips/rtc/pcf8523');

const I2C_BUS = parseInt(process.env.I2C_BUS || '1', 10);
const I2C_ADDR = parseInt(process.env.I2C_ADDR || '0x68', 16);

let passed = 0;
let failed = 0;

function checkTrue(label, condition) {
    if (condition) { console.log('PASS', label); passed++; }
    else { console.log('FAIL', label); failed++; }
}

async function main() {
    const connection = new I2CConnection(I2C_BUS, I2C_ADDR);
    const rtc = new PCF8523Full(connection);
    await rtc.init();
    checkTrue('init', true);

    await rtc.setDatetime(2026, 9, 23, 3, 14, 30, 0);
    const dt = await rtc.getDatetime();
    checkTrue('datetime date', dt.year === 2026 && dt.month === 9 && dt.day === 23 && dt.weekday === 3);
    checkTrue('datetime time', dt.hour === 14 && (dt.minute === 30 || dt.minute === 31));
    checkTrue('oscillator not stopped after setDatetime', !(await rtc.oscillatorStopped()));

    await rtc.setAlarm({ minute: 15, hour: 6 });
    const alarm = await rtc.getAlarm();
    checkTrue('alarm roundtrip', alarm.minute === 15 && alarm.hour === 6 && alarm.day === null && alarm.weekday === null);
    await rtc.setAlarm();

    await rtc.setOffset(-3, 'every_minute');
    const off = await rtc.getOffset();
    checkTrue('offset roundtrip', off.offset === -3 && off.mode === 'every_minute');
    await rtc.setOffset(0);

    // Timer B at 64 Hz from 64 counts should expire in ~1 s.
    await rtc.pollInterrupt();
    await rtc.configureTimerB(64, '64hz');
    let fired = false;
    const deadline = Date.now() + 3000;
    while (Date.now() < deadline) {
        if ((await rtc.pollInterrupt()) & PCF8523Full.SOURCE_TIMER_B) { fired = true; break; }
        await new Promise((r) => setTimeout(r, 50));
    }
    checkTrue('timer b fires', fired);
    await rtc.disableTimerB();

    await connection.close();

    console.log(`===DONE: ${passed} passed, ${failed} failed===`);
    process.exit(failed === 0 ? 0 : 1);
}

main();
